analyses/cell-type-ewings/scripts/utils/clustering-functions.R

```diff
@@ -0,0 +1,212 @@
+# These functions are used to perform clustering, calculate cluster stats,
+# and produce plots related to clusters and cluster stats
+
+# source in jaccard functions
+module_base <- rprojroot::find_root(rprojroot::is_renv_project)
+jaccard_functions <- file.path(module_base, "scripts", "utils", "jaccard-functions.R")
+source(jaccard_functions)
+
+# source in helper functions: plot_density()
+validation_functions <- file.path(module_base, "scripts", "utils", "tumor-validation-helpers.R")
+source(validation_functions)
+
+# Perform clustering -----------------------------------------------------------
+
+# get louvain, jaccard clusters for a specified value of k (nearest neighbors)
+get_clusters <- function(pcs, k) {
+  clusters <- bluster::clusterRows(
+    pcs,
+    bluster::NNGraphParam(
+      k = k,
+      type = "jaccard",
+      cluster.fun = "louvain"
+    )
+  )
+
+  return(clusters)
+}
+
+# define a function to perform clustersweep and get clusters across multiple values of k
```

Comment:
I'd add `(5,40,5)` into this comment since it's hard-coded (which I don't see as a problem necessarily)